### PR TITLE
bluetooth: tester: bap: Add support for BAP/USR tests

### PR DESCRIPTION
**Pull request by mkasenberg**
Main changes:
- Extend Codec capabilities of the server.
- At Codec Config operation initiated autonomously by server, all ASEs with lower IDs than the requested one have to be initiated first due to API limitation.
- In some BAP/USR (server role) test cases the PTS sends the MMI 202 to configure the Sink ASE to streaming state, in some others it does not. For now, let's autonomously send Receiver Start Ready at ASE Enabled.

### CODE REVIEW

File: tests/bluetooth/tester/src/btp_bap.c [hunk 1]
```diff
@@ -524,8 +526,25 @@ static void stream_qos_set(struct bt_bap_stream *stream)
...
 	struct audio_stream *a_stream = CONTAINER_OF(stream, struct audio_stream, stream);
+	struct bt_bap_ep_info info;
+	struct bt_conn_info conn_info;
+	int err;
 
 	LOG_DBG("Enabled stream %p", stream);
+
+	err = bt_bap_ep_get_info(stream->ep, &info);
+	if (err) {
+		LOG_ERR("Could not get EP info for stream %p", stream);
```
Comment (sjanc) [inline tests/bluetooth/tester/src/btp_bap.c:537]: just return here,  if we cannot get ep info we won't be able to properly fill in event data anyway (as info will be left uninitialized).

or maybe send complete event but with failed status?

(same in stream_strated)

File: tests/bluetooth/tester/src/btp_bap.c [hunk 2]
```diff
@@ -524,8 +526,27 @@ static void stream_qos_set(struct bt_bap_stream *stream)
...
+		return;
+	}
+
+	(void)bt_conn_get_info(stream->conn, &conn_info);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_bap.c:541]: Any reason why we are ignoring the return value of `bt_conn_get_info` but not of `bt_bap_ep_get_info`? 

IMO makes sense to treat these 2 the same

File: tests/bluetooth/tester/src/btp_bap.c [hunk 3]
```diff
@@ -1050,7 +1068,11 @@ static uint8_t bap_send(const void *cmd, uint16_t cmd_len,
 	(void)bt_conn_get_info(conn, &conn_info);
 
 	stream = stream_find(audio_conn, cp->ase_id);
-	if (stream == NULL || stream->stream.ep->dir != BT_AUDIO_DIR_SINK) {
+	if (stream == NULL ||
+	    (conn_info.role == BT_HCI_ROLE_CENTRAL &&
+	     stream->stream.ep->dir != BT_AUDIO_DIR_SINK) ||
+	    (conn_info.role == BT_HCI_ROLE_PERIPHERAL &&
+	     stream->stream.ep->dir != BT_AUDIO_DIR_SOURCE)) {
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_bap.c:1075]: The `bt_bap_ep_info` now contains a `can_send` field to make it easier to verify if you can TX on a given stream/ep

File: tests/bluetooth/tester/src/btp_bap.c [hunk 4]
```diff
@@ -1290,6 +1320,81 @@ static bool codec_config_store(struct bt_data *data, void *user_data)
...
+static int client_configure_codec(struct audio_connection *audio_conn, struct bt_conn *conn,
+				  uint8_t ase_id, struct bt_audio_codec_cfg *codec_cfg)
+{
+	int err;
+	struct bt_bap_ep *ep;
+	struct audio_stream *stream;
+
+	stream = stream_find(audio_conn, ase_id);
+
+	if (stream == NULL) {
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_bap.c:1332]:
```suggestion
	stream = stream_find(audio_conn, ase_id);
	if (stream == NULL) {
```

File: tests/bluetooth/tester/src/btp_bap.c [hunk 5]
```diff
@@ -1290,6 +1320,81 @@ static bool codec_config_store(struct bt_data *data, void *user_data)
...
+			return -ENOMEM;
+		}
+
+		if (audio_conn->end_points_count == 0) {
+			return -EINVAL;
+		}
+
+		ep = end_point_find(audio_conn, ase_id);
+
+		if (ep == NULL) {
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_bap.c:1347]:
```suggestion
		ep = end_point_find(audio_conn, ase_id);
		if (ep == NULL) {
```

File: tests/bluetooth/tester/src/btp_bap.c [hunk 6]
```diff
@@ -1290,6 +1320,81 @@ static bool codec_config_store(struct bt_data *data, void *user_data)
...
+
+static int server_configure_codec(struct audio_connection *audio_conn, struct bt_conn *conn,
+				  uint8_t ase_id, struct bt_audio_codec_cfg *codec_cfg)
+{
+	struct audio_stream *stream;
+	int err = 0;
+
+	stream = stream_find(audio_conn, ase_id);
+
+	if (stream == NULL) {
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_bap.c:1368]:
```suggestion
	stream = stream_find(audio_conn, ase_id);
	if (stream == NULL) {
```

**Review by Thalley — DISMISSED** [review]
LGTM with a few suggestions

**Review by jhedberg — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@mkasenberg please rebase.

**Comment by carlescufi** [comment]
@mkasenberg please rebase

**Review by Thalley — APPROVED** [review]

**Review by MariuszSkamra — APPROVED** [review]

**Pull request merged**